lib/adapters/code-action-adapter.ts

```diff
@@ -47,7 +47,7 @@ export default class CodeActionAdapter {
 
     const params = CodeActionAdapter.createCodeActionParams(linterAdapter, editor, range, diagnostics)
     const actions = await connection.codeAction(params)
-    return actions.map((action) => CodeActionAdapter.createCodeAction(action, connection))
+    return (actions || []).map((action) => CodeActionAdapter.createCodeAction(action, connection))
```

Comment:
According to [`codeAction`](https://github.com/atom-community/atom-languageclient/blob/ffbe675bf8a2dc90171a5599dd2dcf88a4398f0f/lib/languageclient.ts#L403) it returns `Promise<Array<lsp.Command | lsp.CodeAction>>` so a promise resolving to `null` or `undefined` is an invalid return type and should probably throw an error instead of being handled the same as an empty array. Or the return type should be updated to include `null` and `undefined` as valid returns.

Reply:
I tend to prefer throwing a valid error instead of failing silently like this. @aminya This is what I was talking about in https://github.com/atom-community/atom-languageclient/pull/132#discussion_r595452569 when I said "types do not catch runtime errors".

Reply:
Thank you for your review!

The spec seems to say that codeAction returns `(Command | CodeAction) [] | null`. Is it not necessary to raise an error if the language server returns valid null?
> https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_codeAction
> Request:
> - method: ‘textDocument/codeAction’
> 
> Response:
> - result: (Command | CodeAction)[] | null

By the way, such a change is also an option.
```diff
 public static async getCodeActions(
   ...
-): Promise<atomIde.CodeAction[]> {
+): Promise<atomIde.CodeAction[] | null> {
   if (linterAdapter == null) {
     return []
   }
   assert(serverCapabilities.codeActionProvider, "Must have the textDocument/codeAction capability")
   const params = CodeActionAdapter.createCodeActionParams(linterAdapter, editor, range, diagnostics)
   const actions = await connection.codeAction(params)
-  return (actions || []).map((action) => CodeActionAdapter.createCodeAction(action, connection))
+  if (actions==null) {return null}
+  return actions.map((action) => CodeActionAdapter.createCodeAction(action, connection))
 }
```

Reply:
I like the early null check rather than still mapping over an empty array.

We should change the return type of [codeAction](https://github.com/atom-community/atom-languageclient/blob/ffbe675bf8a2dc90171a5599dd2dcf88a4398f0f/lib/languageclient.ts#L403) if null is a valid return according to the spec.

Reply:
At https://github.com/atom-community/atom-languageclient/pull/150/commits/e7c26b4f22dbf58ae17788b2a66fa04abeafc466, I changed it to return null instead of an empty array when `actions` value is null.